src/webgpu/shader/execution/evaluation_order.spec.ts

```diff
@@ -0,0 +1,480 @@
+      _body:
+        'let r = (add(&a, 1) == 3) && (a == 3);' + //
+        'return bool_to_i32(r);',
+      _result: 1,
+    },
+    {
+      name: 'RightSE',
+      _body:
+        'let r = (a == 2) && (add(&a, 1) == 3);' + //
+        'return bool_to_i32(r);',
+      _result: 1,
+    },
+    {
+      name: 'LeftmostSE',
+      _body:
+        'let r = (mul3_ret0(&a, &b, &c, 10) == 0) && (a == 20) && (b == 30) && (c == 40);' +
+        'return bool_to_i32(r);',
+      _result: 1,
+    },
+    {
+      name: 'RightmostSE',
+      _body:
+        'let r = (a == 2) && (b == 3) && (c == 4) && (mul3_ret0(&a, &b, &c, 10) == 0);' +
+        'return bool_to_i32(r);',
+      _result: 1,
+    },
+    {
+      name: 'MiddleSE',
+      _body:
+        'let r = (a == 2) && (b == 3) && (mul3_ret0(&a, &b, &c, 10) == 0) && (c == 40);' +
+        'return bool_to_i32(r);',
+      _result: 1,
+    },
+    {
+      name: 'ShortCircuit_And_LhsOnly',
+      _body:
+        // rhs should not execute
+        'let t = (a != 2) && (mul(&a, 10) == 20);' + //
+        'let r = (a == 2);' +
+        'return bool_to_i32(r);',
+      _result: 1,
+    },
+    {
+      name: 'ShortCircuit_And_LhsAndRhs',
+      _body:
+        // rhs should execute
+        'let t = (a == 2) && (mul(&a, 10) == 20);' + //
+        'let r = (a == 20);' +
+        'return bool_to_i32(r);',
+      _result: 1,
+    },
+    {
+      name: 'ShortCircuit_Or_LhsOnly',
+      _body:
+        // rhs should not execute
+        'let t = (a == 2) || (mul(&a, 10) == 20);' + //
+        'let r = (a == 2);' +
+        'return bool_to_i32(r);',
+      _result: 1,
+    },
+    {
+      name: 'ShortCircuit_Or_LhsAndRhs',
+      _body:
+        // rhs should execute
+        'let t = (a != 2) || (mul(&a, 10) == 20);' + //
+        'let r = (a == 20);' +
+        'return bool_to_i32(r);',
+      _result: 1,
+    },
+  ])
+  .fn(t => run(t, t.params._body, t.params._result));
+
+g.test('binary_mixed')
+  .specURL('https://gpuweb.github.io/gpuweb/wgsl/#arithmetic-expr')
+  .specURL('https://gpuweb.github.io/gpuweb/wgsl/#logical-expr')
+  .desc('Tests order of evaluation of both arithmetic and logical binary expressions.')
+  .paramsSimple([
+    {
+      name: 'ArithAndLogical',
+      _body: 'return mul(&a, 10) - bool_to_i32(mul(&a, 10) == 200 && mul(&a, 10) == 2000);',
+      _result: 20 - 1,
+    },
+    {
+      name: 'LogicalAndArith',
+      _body: 'return bool_to_i32(mul(&a, 10) == 20 && mul(&a, 10) == 200) - mul(&a, 10);',
+      _result: 1 - 2000,
+    },
+  ])
+  .fn(t => run(t, t.params._body, t.params._result));
+
+g.test('call')
+  .specURL('https://gpuweb.github.io/gpuweb/wgsl/#function-calls')
+  .desc('Tests order of evaluation of call expressions.')
+  .paramsSimple([
+    {
+      name: 'OneSE', //
+      _body: 'return sub3(mul(&a, 10), a, 3);',
+      _result: 20 - 20 - 3,
+    },
+    {
+      name: 'AllSE',
+      _body: 'return sub3(mul(&a, 10), mul(&a, 10), mul(&a, 10));',
```

Comment:
I think this test would be stronger if the `mul()` calls were all different in some way. As they are now you can execute them in any order and still get the same answer.

The same applies to a handful of other tests.

Reply:
> As they are now you can execute them in any order and still get the same answer.
Almost true, but yes. Of the 6 combinations of possible execution orders, there are three possible answers, so that's no good. I changed it to a "subtraction of multiplications":

```
fn sub_mul(a : i32, a_mul : i32, b : i32, b_mul : i32, c : i32, c_mul : i32) -> i32 {
  return a * a_mul - b * b_mul - c * c_mul;
}
```

And replaced with:
```
'return sub_mul(mul(&a, 2), 2, mul(&a, 2), 3, mul(&a, 2), 4);'
```

Which gives 6 different answers for all 6 six combinations.

Good?